### ISSUE TITLE
pageVisibilityTracker: Issue with Custom metric " Page Visible Time"

### ISSUE THREAD

**Issue by EW0JY** [googleanalytics/autotrack#205]
Hi!

I'm using the following setup of the plugin:

`ga('require', 'pageVisibilityTracker', {
  sendInitialPageview: true,
});
// The command below is no longer needed.
// ga('send', 'pageview');`

After reading the documentation many times over, I'm still unclear on the following:

- Does the above code suffice to send the visible page time, or is there additional code necessary?

The reason I'm confused is because the documentation says the [default values](https://github.com/googleanalytics/autotrack/blob/master/docs/plugins/page-visibility-tracker.md#page-visibility-events) already send the elapsed time as the "eventValue".

But then I read this a bit further up in the doc: 

> to track the time a page was visible is to create a custom metric called Page Visible Time that you set in your plugin configuration options

So basically, my question is:

- Do I only need to set up a custom metric within Google Analytics, or alter the above mentioned plugin code that I'm currently using?

Any help would be greatly appreciated!

Thank you!
Flo

**Comment by EW0JY** [comment]
Thanks so much, Philip!

That helped a lot.

**Comment by philipwalton** [comment]
> if I didn't set up any custom metric in Analytics, where could I even see the page view time that gets sent on the event value?

Go to "Behavior" > "Events" > "Top Events" and then click on "Page Visibility". There you should see both the total value and the average value.

> Would that work like that?

I use time, here's my config:

![screen shot 2017-10-17 at 11 36 26 am](https://user-images.githubusercontent.com/326742/31682611-7e68f71e-b32f-11e7-9e3d-881d00f51f0b.png)

> I guess both questions come down to the fact that I believe the custom metric serves as some sort of mapping for the value received. That's why I'm unsure where it gets mapped to if I didn't set it up.

It's not really a mapping, it's just a user-defined metric. It might help to think of it like a traditional metric like pageviews, but one that you define and implement yourself (you could actually implement pageviews yourself with a custom metric, I describe a similar strategy for implementing page load tracking via custom metrics in [this article](https://philipwalton.com/articles/how-we-track-pageviews-is-all-wrong/)).

You can read more about custom metrics here:
https://support.google.com/analytics/answer/2709828?hl=en

**Comment by EW0JY** [comment]
Again, thanks so much!

I set up the custom metric "Page Visible Time" exactly llike you said.
Checking the data within Analytics today, I can see under events that data is getting through, showing as "Page Visibility" categorie and the eventValues are also filled with data (which must be the total of all measured seconds).

Here's the issue:
When I use that custom metric in a custom report, it doesn't show any data, it's all 0.
What could be the reason for this?

**Comment by philipwalton** [comment]
Did you set the [`visibleMetricIndex`](https://github.com/googleanalytics/autotrack/blob/master/docs/plugins/page-visibility-tracker.md#options) option? (Sorry, it may have been misleading when I said above that you don't need to change your code).

It should look something like this:

```js
ga('require', 'pageVisibilityTracker', {
  sendInitialPageview: true,
  visibleMetricIndex: X // Where 'X' is whatever the index is in GA
});
```

Setting this options tells the plugin to send the custom metric field `metricX` with the hit, and use the same value as `eventValue`.

You can read more about how custom dimensions and metrics work with analytics.js here:
https://developers.google.com/analytics/devguides/collection/analyticsjs/custom-dims-mets

**Comment by EW0JY** [comment]
Alright, so the code looks like this now:

`ga('require', 'pageVisibilityTracker', {
  sendInitialPageview: true,
  visibleMetricIndex: 1,
});`

So from now on the custom report should work, i.e. the custom metric "Page Visible Time" should be "filled" with data?

What I don't get is this: as I said, even now (before I updated the code with the visibleMetricIndex) the "Page Visibility" category shows plenty "eventValue", so it seems that there was already some data exchange happening.
- What exactly is that current "event value", if it shouldn't be there since the MetricIndex wasn't set yet?
- Will that interfere with the "new" data that will come in soon?

**Comment by philipwalton** [comment]
I think this idea of "data exchange" is what's causing your confusion. The `eventValue` and `metric1` fields are two completely separate metrics. The plugin is just setting them both to the same value.

Before you added `visibleMetricIndex: 1`, the data you were sending to Google Analytics looked something like this:

```
v=1&t=event&tid=UA-XXXX-Y&cid=uuid&ec=Page%20Visibility&ev=123
```

And after adding `visibleMetricIndex: 1`, it looks like this:

```
v=1&t=event&tid=UA-XXXX-Y&cid=uuid&ec=Page%20Visibility&ev=123&cm1=123
```

(notice the `cm1=123` added to the end)

The reason `eventValue` showed up before is because the `eventValue` metric is a standard metric and its the same for every GA user. Custom metrics, on the other hand, are unique to each user, so autotrack can't send data for them until it knows the index.

**Comment by EW0JY** [comment]
ok, so if I understand you correctly, after setting the `visibleMetricIndex: 1`, I should be able to use the custom metric that I had already set up according to your suggestions?
What I don't quite understand is how that can work without me telling Google Analytics which index (in this case "1") it should use to fill the custom metric. But that doesn't matter, as long as it works :)

**Comment by philipwalton** [comment]
> What I don't quite understand is how that can work without me telling Google Analytics which index (in this case "1") it should use to fill the custom metric. But that doesn't matter, as long as it works :)

When you created the metric, Google Analytics automatically assigns it an index. As long as you're using that index, it should work.

![screen shot 2017-10-18 at 3 14 16 pm](https://user-images.githubusercontent.com/326742/31745471-a40b5636-b417-11e7-9f15-c345f08df231.png)

**Comment by EW0JY** [comment]
Alright, that makes a lot of sense now, thanks for pointing that our :) !

Btw: does your plugin fix the issue regarding measuring the time if a user only ever visits one page and then leaves?
With the standard GA code this results in 0 seconds duration on every exit page.

**Comment by philipwalton** [comment]
>  Does your plugin fix the issue regarding measuring the time if a user only ever visits one page and then leaves? With the standard GA code this results in 0 seconds duration on every exit page.

Yep, that was the primary motivation for making the plugin :)

Note that it won't log any time if the user visit one page, leaves the page visible, and then the session times out, but that's much less common. If the user closes the tab or switches tabs, time will be logged.
